[inductor] Add ProcessPool-in-subproc worker_start_method

jansel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #125451
* __->__ #122941
* #124682

In many environments using fork-based parallelism causes issues because user processes are not fork-safe.  This moves our parallel compile work pool into a subprocess that we control (that should be fork safe).

Perf run: https://github.com/pytorch/pytorch/actions/runs/8486887873



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

Differential Revision: [D57142035](https://our.internmc.facebook.com/intern/diff/D57142035)

jansel committed:
Update

[ghstack-poisoned]

Skylion007 commented on torch/_inductor/compile_worker/subproc_pool.py:
```diff
@@ -0,0 +1,238 @@
+                self.write_pipe.close()
+            self.process.wait(300)
+        except OSError:
+            pass
```
probably should log this exception in debug mode, no?

jansel committed:
Update

[ghstack-poisoned]

jansel committed:
Update

[ghstack-poisoned]

jansel committed:
Update

[ghstack-poisoned]

jansel committed:
Update

[ghstack-poisoned]

jansel committed:
Update

[ghstack-poisoned]

jansel committed:
Update

[ghstack-poisoned]

jansel committed:
Update

[ghstack-poisoned]

jansel committed:
Update

[ghstack-poisoned]

jansel committed:
Update

[ghstack-poisoned]

jansel committed:
Update

[ghstack-poisoned]

jansel committed:
Update

[ghstack-poisoned]

jansel committed:
Update

[ghstack-poisoned]

jansel committed:
Update

[ghstack-poisoned]

jansel committed:
Update

[ghstack-poisoned]

jansel committed:
Update

[ghstack-poisoned]

jansel committed:
Update

[ghstack-poisoned]

jansel committed:
Update

[ghstack-poisoned]

jansel committed:
Update

[ghstack-poisoned]

jansel committed:
Update

[ghstack-poisoned]

jansel committed:
Update

[ghstack-poisoned]

jansel committed:
Update

[ghstack-poisoned]

jansel committed:
Update

[ghstack-poisoned]

masnesral committed:
Update on "[inductor] Add ProcessPool-in-subproc worker_start_method"


In many environments using fork-based parallelism causes issues because user processes are not fork-safe.  This moves our parallel compile work pool into a subprocess that we control (that should be fork safe).

Perf run: https://github.com/pytorch/pytorch/actions/runs/8486887873



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

masnesral committed:
Update on "[inductor] Add ProcessPool-in-subproc worker_start_method"


In many environments using fork-based parallelism causes issues because user processes are not fork-safe.  This moves our parallel compile work pool into a subprocess that we control (that should be fork safe).

Perf run: https://github.com/pytorch/pytorch/actions/runs/8486887873



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

masnesral committed:
Update on "[inductor] Add ProcessPool-in-subproc worker_start_method"


In many environments using fork-based parallelism causes issues because user processes are not fork-safe.  This moves our parallel compile work pool into a subprocess that we control (that should be fork safe).

Perf run: https://github.com/pytorch/pytorch/actions/runs/8486887873



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

masnesral commented:
@masnesral has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D57142035).
